docs/DesignDocs/DeepCopyOutParameters.md

```diff
@@ -0,0 +1,718 @@
+    +       if (!_deepcopy_out_buffer)
+    +       {
+    +           _result = OE_OUT_OF_MEMORY;
+    +           goto done;
+    +       }
+    +   }
+
+    +   /* Serialize the deep-copied content into the buffer. */
+    +   if (pargs_in->ptr && pargs_in->ptr->blob_array)
+    +       OE_WRITE_DEEPCOPY_OUT_PARAM(pargs_in->ptr->blob_array, ((size_t)pargs_in->ptr->num * sizeof(Blob)));
+    +   for (size_t _i_2 = 0; _i_2 < pargs_in->ptr->num; _i_2++)
+    +   {
+    +       if (pargs_in->ptr && pargs_in->ptr->blob_array && pargs_in->ptr->blob_array[_i_2].buf)
+    +           OE_WRITE_DEEPCOPY_OUT_PARAM(pargs_in->ptr->blob_array[_i_2].buf, pargs_in->ptr->blob_array[_i_2].len);
+    +   }
+
+    +   if (_deepcopy_out_buffer_offset != _deepcopy_out_buffer_size)
+    +   {
+    +       _result = OE_FAILURE;
+    +       goto done;
+    +   }
+    +   OE_ADD_SIZE(output_buffer_offset, _deepcopy_out_buffer_size);
+
+    +   /* Set the _deepcopy_out_buffer and _deepcopy_out_buffer as part of pargs_out. */
+    +   pargs_out->deepcopy_out_buffer = _deepcopy_out_buffer;
+    +   pargs_out->deepcopy_out_buffer_size = _deepcopy_out_buffer_size;
+
+        /* Propagate errno back to enclave. */
+        /* Errno propagation not enabled. */
+
+        /* Success. */
+        _result = OE_OK;
+        *output_bytes_written = output_buffer_offset;
+
+    done:
+    +   /* Free pargs_out->deepcopy_out_buffer on failure. */
+    +   if (_result != OE_OK)
+    +   {
+    +       free(pargs_out->deepcopy_out_buffer);
+    +       pargs_out->deepcopy_out_buffer = NULL;
+    +       pargs_out->deepcopy_out_buffer_size = 0;
+    +   }
+
+    +   /* Free nested buffers allocated by the host. */
+    +   if (pargs_in->ptr)
+    +   {
+    +       for (size_t _i_1 = 0; _i_1 < 1; _i_1++)
+    +       {
+    +           if (pargs_in->ptr[_i_1].blob_array)
+    +           {
+    +               for (size_t _i_2 = 0; _i_2 < pargs_in->ptr[_i_1].num; _i_2++)
+    +               {
+    +                   free(pargs_in->ptr[_i_1].blob_array[_i_2].buf);
+    +                   pargs_in->ptr[_i_1].blob_array[_i_2].buf = NULL;
+    +               }
+    +           }
+    +           free(pargs_in->ptr[_i_1].blob_array);
+    +           pargs_in->ptr[_i_1].blob_array = NULL;
+    +       }
+    +   }
+
+        if (pargs_out && output_buffer_size >= sizeof(*pargs_out))
+            pargs_out->_result = _result;
+    }
+    ```
+
+Conclusion and Plan
+-----------
+
+This document presents the design, which supports deep-copy `out` parameter.
+Moving forward, we plan to support the design as an experimental feature first.
+That is, users need to explicitly opt-in the feature (i.e., specifying `--experimental` flag in the
+command line when invoking the oeedger8r tool). Doing so allows OE to support the deep-copy `out` parameter
+in SGX while not introducing breaking changes to the OE's support on the OP-TEE.
```

Comment:
```suggestion
in SGX while not introducing breaking changes to the OE's support on the OP-TEE.
```
@HernanGatta is there anything in here that would break something that works on OP-TEE?  If not, then this doesn't need to be experimental.

Reply:
The design won't break existing OP-TEE ECALLs/OCALLs that don't use the deep-copy out annotations (there shouldn't be any).
Even if developers try to use deep-copy out on OP-TEE, the compilation would go thru but the code just won't work unless we implement the OP-TEE support internally.

Given that we currently don't specify target TEE to the oeedger8r tool, we don't have a good way to tell developers that deep-copy out is not supported on OP-TEE. As a result, the current plan is guarding the feature with the `--experimental` flag, which makes the feature implicit. If the flag is not set, the developer will see an error message like:
```
`out' for user defined type `MyStruct\\*' requires the --experimental option`.
```

@dthaler @anakrish Do you think we can make this as non-experimental? We do have the switchless call feature that is similar to this case (only works for SGX).

Alternatively, we could introduce a TEE flag to oeedger8r or let the tool to detect the target based on the platform (x86 or arm). We could also add the OP-TEE support, which requires @HernanGatta's inputs. I think both of them could be future work.

Reply:
Let's first merge this in as experimental, and get feedback from the user that wants this.

Reply:
Unless we know of things that would break OP-TEE, I don't think OP-TEE should be any reason to make this be experimental.  If we think it needs to be experimental for SGX because SGX might break, then ok.  But I'd probably argue for non-experimental.

Reply:
We're thinking of following the previous path on deep-copy support: making the feature as experimental first and then get feedback from users/ask for security reviews before making this as non-experimental.

Reply:
I agree with what Ming-Wei said. Let’s get feedback and see if some folks can look at the security implications. Cc @mrragava